ada_feeding/ada_feeding/idioms/retry_call_ros_service.py

```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+This module defines the retry_call_ros_service idiom, which will call a ROS
+service, optionally check the response value, and retry up to a pre-specified
+number of times if there is a failure.
+"""
+
+# Standard imports
+import logging
+from typing import Any, List, Optional
+
+# Third-party imports
+import py_trees
+from py_trees.blackboard import Blackboard
+from py_trees_ros.service_clients import FromBlackboard, FromConstant
+
+
+def retry_call_ros_service(
+    name: str,
+    service_type: Any,
+    service_name: str,
+    key_request: Optional[str] = None,
+    request: Optional[Any] = None,
+    key_response: Optional[str] = None,
+    response_checks: Optional[List[py_trees.common.ComparisonExpression]] = None,
+    max_retries: int = 3,
```

Comment:
Is there a reason for max_retries to be 3 here? Would we want to change it in the future?

Reply:
If not, you can ignore this comment.

Reply:
I just chose 3 because it seemed reasonable to me.

Any behavior tree that uses this idiom can pass a custom parameter to this function for a different value for max_retries. For example, we can change the code of `pre_moveto_config.py` to change the number of retries for each of its service calls. 

A question is whether we would want to expose that as a parameter to the behavior tree itself, which can then be defined via the yaml file. The benefit of that is that we can change the num retries without changing the code. But I'm not convinced that is necessary, because the main reason to retry on a service call is if the service isn't available, and beyond some number of retries it won't make sense to keep retrying. So I think we can probably hardcode a reasonable value per service call, and therefore don't need to expose the parameter all the way up to the level of the yaml file.